packages/animations/lib/src/shared_z_axis_transition.dart

```diff
@@ -0,0 +1,394 @@
+///       items: const <BottomNavigationBarItem>[
+///         BottomNavigationBarItem(
+///           icon: Icon(Icons.home),
+///           title: Text('White'),
+///         ),
+///         BottomNavigationBarItem(
+///           icon: Icon(Icons.business),
+///           title: Text('Red'),
+///         ),
+///         BottomNavigationBarItem(
+///           icon: Icon(Icons.school),
+///           title: Text('Yellow'),
+///         ),
+///       ],
+///       currentIndex: _selectedIndex,
+///       onTap: (int index) {
+///         setState(() {
+///           _selectedIndex = index;
+///         });
+///       },
+///     ),
+///   );
+/// }
+/// ```
+class SharedZAxisTransition extends StatefulWidget {
+  /// Creates a [SharedZAxisTransition].
+  ///
+  /// The [animation] and [secondaryAnimation] argument are required and must
+  /// not be null.
+  const SharedZAxisTransition({
+    Key key,
+    @required this.animation,
+    @required this.secondaryAnimation,
+    this.child,
+  }) : super(key: key);
+
+  /// The animation that drives the [child]'s entrance and exit.
+  ///
+  /// See also:
+  ///
+  ///  * [TransitionRoute.animate], which is the value given to this property
+  ///    when it is used as a page transition.
+  final Animation<double> animation;
+
+  /// The animation that transitions [child] when new content is pushed on top
+  /// of it.
+  ///
+  /// See also:
+  ///
+  ///  * [TransitionRoute.secondaryAnimation], which is the value given to this
+  ///    property when the it is used as a page transition.
+  final Animation<double> secondaryAnimation;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// This widget will transition in and out as driven by [animation] and
+  /// [secondaryAnimation].
+  final Widget child;
+
+  @override
+  _SharedZAxisTransitionState createState() => _SharedZAxisTransitionState();
+}
+
+class _SharedZAxisTransitionState extends State<SharedZAxisTransition> {
+  AnimationStatus _effectiveAnimationStatus;
+  AnimationStatus _effectiveSecondaryAnimationStatus;
+
+  @override
+  void initState() {
+    super.initState();
+    _effectiveAnimationStatus = widget.animation.status;
+    _effectiveSecondaryAnimationStatus = widget.secondaryAnimation.status;
+    widget.animation.addStatusListener(_animationListener);
+    widget.secondaryAnimation.addStatusListener(_secondaryAnimationListener);
+  }
+
+  void _animationListener(AnimationStatus animationStatus) {
+    _effectiveAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveAnimationStatus,
+      current: animationStatus,
+    );
+  }
+
+  void _secondaryAnimationListener(AnimationStatus animationStatus) {
+    _effectiveSecondaryAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveSecondaryAnimationStatus,
+      current: animationStatus,
+    );
+  }
+
+  // When a transition is interrupted midway we just want to play the ongoing
```

Comment:
NICE explanation

Reply:
Credit would go to @goderbauer for this one :)